lib/Parser/Parser.yy

```diff
@@ -50,155 +50,544 @@
+      /* empty */
+    | template_definition
+    ;
+
+template_definition:
+      ltOp template_parameter_list gtOp
+    ;
+
+template_parameter_list:
+      template_parameter
+    | template_parameter_list comma template_parameter
+    ;
+
+template_parameter:
+      ident
+    ;
+
+struct_body:
+      lBrace struct_field_list rBrace
+    ;
+
+struct_field_list:
+      struct_field
+    | struct_field_list comma struct_field
+    ;
+
+struct_field:
+      ident colon type
+    | ident colon type equal expression
+    ;
+
+enum_declaration:
+      attributes enumKw ident colon type enum_body
+    { std::cerr << "Parsed enum declaration" << std::endl; }
+    ;
+
+enum_body:
+      lBrace enum_variant_list rBrace
+    ;
+
+enum_variant_list:
+      enum_variant
+    | enum_variant_list comma enum_variant
+    ;
+
+enum_variant:
+      ident
+    | ident equal expression
     ;
 
-declarations
-    : declaration
-    | declarations declaration
+typealias_declaration:
+      attributes typealiasKw ident template_definition_opt equal type semi
+    { std::cerr << "Parsed typealias declaration" << std::endl; }
     ;
 
-declaration
-    : function_decl
-    | var_decl
-    | struct_decl
+/*--------------------------------*/
+/* Function declarations          */
+/*--------------------------------*/
+
+function_declaration:
+      attributes funcKw ident template_definition_opt function_signature function_body
+    { std::cerr << "Parsed function declaration" << std::endl; }
     ;
 
-function_decl
-    : funcKw ident lParen rParen compound_stmt
-    { std::cout << "Parsed " << ($2.getLexeme().str()) << " function declaration\n"; }
+function_signature:
+      lParen parameter_list_opt rParen arrow type
+    | lParen parameter_list_opt rParen
     ;
 
-var_decl
-    : letKw ident semi
-    | varKw ident semi
-    { std::cout << "Parsed variable declaration\n"; }
+parameter_list_opt:
+      /* empty */
+    | parameter_list
     ;
 
-struct_decl
-    : structKw ident lBrace struct_members rBrace
-    { std::cout << "Parsed struct declaration\n"; }
+parameter_list:
+      parameter
+    | parameter_list comma parameter
     ;
 
-struct_members
-    : /* empty */
-    | struct_members struct_member
+parameter:
+      ident colon type
+    | ident colon type equal expression
     ;
 
-struct_member
-    : var_decl
+function_body:
+      block
+    | semi
     ;
 
-compound_stmt
-    : lBrace statements rBrace
+block:
+      lBrace statement_list rBrace
     ;
 
-statements
-    : /* empty */
-    | statements statement
+statement_list:
+      /* empty */
+    | statement_list statement
     ;
 
-statement
-    : expr_stmt
-    | compound_stmt
+statement:
+      block
+    | expression semi
+    | var_stmt
+    | let_stmt
+    | return_stmt
     | if_stmt
     | while_stmt
-    | return_stmt
+    | for_stmt
+    | break_stmt
+    | continue_stmt
+    | assignment_stmt
+    | semi
+    ;
+
+var_stmt:
+      varKw ident type_opt equal_opt expression_opt semi
```

Comment:
here both equal and expression are optional, allowing weird things:
```
var hello: str =;
var hello: str 0;
var hello 0;
var hello =;
```
are all allowed by your grammar, but make no sense. Either we have both equal and expression, or none. (type_opt is correct as it's either empty or both colon and type)